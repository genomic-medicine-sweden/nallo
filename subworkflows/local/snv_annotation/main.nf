include { ANNOTATE_CADD                               } from '../annotate_cadd/main'
include { ECHTVAR_ANNO                                } from '../../../modules/local/echtvar/anno/main'
include { BCFTOOLS_FILLTAGS as BCFTOOLS_FILLTAGS_ANNO } from '../../../modules/local/bcftools/filltags/main'
include { ENSEMBLVEP_VEP                              } from '../../../modules/nf-core/ensemblvep/vep/main'
include { TABIX_TABIX as TABIX_VEP                    } from '../../../modules/nf-core/tabix/tabix/main'

workflow SNV_ANNOTATION {

    take:
    ch_vcf                // channel [mandatory] [ val(meta), path(vcf) ]
    ch_databases          // channel: [mandatory] [ val(meta), path(db) ]
    ch_fasta              // channel: [mandatory] [ val(meta), path(fasta) ]
    ch_fai                // channel: [mandatory] [ val(meta), path(fai) ]
    ch_vep_cache          // channel: [mandatory] [ path(cache) ]
    val_vep_cache_version // string: [mandatory] default: 110
    ch_vep_extra_files    // channel: [mandatory] [ path(files) ]
    val_annotate_cadd     // bool: [mandatory]
    ch_cadd_header        // channel: [mandatory] [ path(txt) ]
    ch_cadd_resources     // channel: [mandatory] [ path(annotation) ]
    ch_cadd_prescored     // channel: [mandatory] [ path(prescored) ]

    main:
    ch_versions = Channel.empty()
    ch_vep_in   = Channel.empty()

    // Annotate with chosen databases (GNOMAD,CADD + SAMPLES_DB)
    ECHTVAR_ANNO ( ch_vcf, ch_databases )
    ch_versions = ch_versions.mix(ECHTVAR_ANNO.out.versions)

    BCFTOOLS_FILLTAGS_ANNO(ECHTVAR_ANNO.out.bcf)
    ch_versions = ch_versions.mix(BCFTOOLS_FILLTAGS_ANNO.out.versions)

    // Annotating with CADD
    if (val_annotate_cadd) {
        ANNOTATE_CADD (
            ch_fai,
            BCFTOOLS_FILLTAGS_ANNO.out.vcf,
            BCFTOOLS_FILLTAGS_ANNO.out.tbi,
            ch_cadd_header,
            ch_cadd_resources,
            ch_cadd_prescored
        )
        ch_versions = ch_versions.mix(ANNOTATE_CADD.out.versions)

        ANNOTATE_CADD.out.vcf
            .map { meta, vcf -> [ meta, vcf, [] ] }
            .set { ch_vep_in }

    } else {
        BCFTOOLS_FILLTAGS_ANNO.out.vcf
            .map { meta, vcf -> [ meta, vcf, [] ] }
            .set { ch_vep_in }

    }

    ENSEMBLVEP_VEP (
        ch_vep_in,
        "GRCh38",
        "homo_sapiens",
        val_vep_cache_version,
        ch_vep_cache,
        ch_fasta,
        ch_vep_extra_files
    )
    ch_versions = ch_versions.mix(ENSEMBLVEP_VEP.out.versions)

    TABIX_VEP ( ENSEMBLVEP_VEP.out.vcf )
    ch_versions = ch_versions.mix(TABIX_VEP.out.versions)

    emit:
    vcf      = ENSEMBLVEP_VEP.out.vcf
    tbi      = TABIX_VEP.out.tbi
    versions = ch_versions
}
