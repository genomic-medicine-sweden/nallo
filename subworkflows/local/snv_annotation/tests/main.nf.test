nextflow_workflow {

    name "Test Workflow SNV_ANNOTATION"
    script "../"
    workflow "SNV_ANNOTATION"
    config "./nextflow.config"

    setup {
        run("GUNZIP") {
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = [
                    [ id:'hg38' ],
                    file(params.pipelines_testdata_base_path + 'nallo/reference/hg38.test.fa.gz', checkIfExists: true)
                ]
                """
            }
        }
        run("SAMTOOLS_FAIDX") {
            script "../../../../modules/nf-core/samtools/faidx/main.nf"
            process {
                """
                input[0] = GUNZIP.out.gunzip
                input[1] = [[],[]]
                """
            }
        }
        run("MINIMAP2_ALIGN") {
            script "../../../../modules/nf-core/minimap2/align/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'test', num_intervals:1 ],
                    file(params.pipelines_testdata_base_path + 'nallo/testdata/HG002_PacBio_Revio.fastq.gz', checkIfExists: true)
                ]
                input[1] = GUNZIP.out.gunzip
                input[2] = true
                input[3] = 'csi'
                input[4] = false
                input[5] = false
                """
            }
        }
        run("SHORT_VARIANT_CALLING") {
            script "../../short_variant_calling/main.nf"
            process {
                """
                input[0] = MINIMAP2_ALIGN.out.bam
                            .join(MINIMAP2_ALIGN.out.index)
                            .join(Channel.of([
                                [ id: 'test', num_intervals:1 ],
                                file(params.pipelines_testdata_base_path + 'nallo/reference/test_data.bed', checkifexists: true)
                            ]))
                input[1] = GUNZIP.out.gunzip
                input[2] = SAMTOOLS_FAIDX.out.fai
                input[3] = [
                    [],[]
                ]
                input[4] = [
                    [],[]
                ]
            """
            }
        }
        run("UNTAR") {
            script "../../../../modules/nf-core/untar/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'vep_cache' ],
                    file(params.pipelines_testdata_base_path + 'nallo/reference/vep_cache_test_data.tar.gz', checkIfExists:true)
                ]
                """
            }
        }

    }

    test("bcf, db, vep_cache, '110'") {

        when {
            workflow {
                """
                input[0] = SHORT_VARIANT_CALLING.out.combined_bcf
                input[1] = [
                    file(params.pipelines_testdata_base_path + 'nallo/reference/cadd.v1.6.hg38.test_data.zip', checkIfExists: true)
                ]
                input[2] = GUNZIP.out.gunzip
                input[3] = SAMTOOLS_FAIDX.out.fai
                input[4] = UNTAR.out.untar.map { meta, cache -> cache }
                input[5] = Channel.value('110')
                input[6] = [
                    file(params.pipelines_testdata_base_path + 'nallo/reference/vep_plugin_files.csv', checkIfExists: true)
                ]
                input[7] = false
                input[8] = Channel.value([])
                input[9] = null
                input[10] = null
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    file(workflow.out.tbi.get(0).get(1)).name,
                    path(workflow.out.vcf.get(0).get(1)).linesGzip[48..118], // Stable
                    path(workflow.out.vcf.get(0).get(1)).linesGzip[121..150] // Stable
                ).match() }
            )
        }

    }

    test("bcf, db, vep_cache, '110', -stub") {

        options "-stub"

        when {
            params {
            }
            workflow {
                """
                input[0] = SHORT_VARIANT_CALLING.out.combined_bcf
                input[1] = [
                    file(params.pipelines_testdata_base_path + 'nallo/reference/cadd.v1.6.hg38.test_data.zip', checkIfExists: true)
                ]
                input[2] = GUNZIP.out.gunzip
                input[3] = SAMTOOLS_FAIDX.out.fai
                input[4] = UNTAR.out.untar.map { meta, cache -> cache }
                input[5] = Channel.value('110')
                input[6] = [
                    file(params.pipelines_testdata_base_path + 'nallo/reference/vep_plugin_files.csv', checkIfExists: true)
                ]
                input[7] = false
                input[8] = Channel.value([])
                input[9] = null
                input[10] = null
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    workflow.out.tbi,
                    workflow.out.vcf
                ).match() }
            )
        }

    }

}
