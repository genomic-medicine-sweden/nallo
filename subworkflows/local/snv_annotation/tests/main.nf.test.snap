{
    "bcf, db, vep_cache, '110', -stub": {
        "content": [
            [
                "versions.yml:md5,1c67249781a135b3abbc3c7e63af4d0b",
                "versions.yml:md5,503122d4650f6a8a39b4e810063d6c22",
                "versions.yml:md5,c0e55e36a31ed71acf25702b7d059533",
                "versions.yml:md5,f533675b4401f25d7dcfc9f63abeb619"
            ],
            [
                [
                    {
                        "id": "test_data.bed",
                        "project": null,
                        "contains_affected": false
                    },
                    "test_data.bed.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                [
                    {
                        "id": "test_data.bed",
                        "project": null,
                        "contains_affected": false
                    },
                    "test_data.bed.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-07T14:29:17.858999735"
    },
    "bcf, db, vep_cache, '110'": {
        "content": [
            [
                "versions.yml:md5,1c67249781a135b3abbc3c7e63af4d0b",
                "versions.yml:md5,503122d4650f6a8a39b4e810063d6c22",
                "versions.yml:md5,c0e55e36a31ed71acf25702b7d059533",
                "versions.yml:md5,f533675b4401f25d7dcfc9f63abeb619"
            ],
            "test_data.bed.vcf.gz.tbi",
            "ad2f3def07bee1f4bab9cfe456dada3e",
            "VcfFile [chromosomes=[chrX, chr16], sampleCount=1, variantCount=121, phased=false, phasedAutodetect=false]"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-08T11:44:29.495925133"
    }
}
